### ISSUE TITLE
Displaying Thumbnails which are in a subfolder of the main gcode folder

### ISSUE THREAD

**Issue by davisgoodman** [NonaSuomy/Moonraker-Home-Assistant#1]
I've used your code for a bit but had issues with the layer count and also with the Thumbnails. I found out that if the gcode file is in the main folder the metadata for the thumbnail is given as the relative path against the gcode file itself. So if a file in the gcodes/subfolder/myfile.gcode, the relative path of the thumbnail will be .thumbs/myfile.png but and won't display as it should be subfolder/.thumbs/myfile.png. 

I did a test and changed the state of the sensor.3d_printer_object_thumbnails in the developper tool to add the subfolder in front the state and then the thumbnail displayed in the card.

Is there anyway to make sure the subfolder are included in the path of the thumbnails? 

Maybe I should ask the Moonraker team

**Comment by davisgoodman** [comment]
Discussed with the moonraker team and we need to be able in the templates to add the subdirectory if it exists in front of  the thumbnail's relative path. I'm not sure how to do this. Not proficient enough in jinja2 templating

**Comment by NonaSuomy** [comment]
Is the layer count correct in moonraker? As it’s just pulling data from there via json. What is determining where the gcode goes for you? The thumbnails graphic url is reported from the entity sensor.3d_printer_object_thumbnails when I look at the json nfo it already adds .thumbs/ to the front of the thumbnail for me? You can make a manual card and populate it with all the 3d printer sensor values you will see that .thumbs is there? (I only upload gcode with SuperSlicer)
![E5A13B14-F36B-4922-8F07-33304891517D](https://user-images.githubusercontent.com/1906575/216603383-e907fca6-6420-4270-9542-d947831b6073.jpeg)

**Comment by davisgoodman** [comment]
For the thumbnails the issue  will show up if you upload your gcode file in a subdirectory in Mainsail or Fluid

I suppose it depends which slicer is used but with SuperSlicer you are able to upload into a particular subdirectory say you have a particular project and you want to regroup your print files.

 

In that particular case the print filename will show the subdirectory in the file path but not the thumbnail.

 

A typical file uploaded to Mainsail within a subfolder would have a path of  “subfolder/filename.gcode but the path for the thumbnail would only be .thumbs/filename.gcode which isn’t the correct path which should be “subdirectory/.thumbs/filename.code. 

With the help of someone on the Home Assistant Discord server, I modified some of the sensors to add the subdirectory to the thumbnail path.

Works great

 
``` yaml

    - name: 3d_printer_object_thumbnails_path

      unique_id: "192.168.168.293988e9be-c910-4634-8205-38d53170d5a6"

      state: '{{ states.sensor.printer_3d_file_metadata.attributes["thumbnails"][1]["relative_path"] }}' 

      attributes:

        friendly_name: "Object Thumbnails Path"

        

    - name: 3d_printer_thumbnail

      unique_id: "192.168.168.293988e9be-c910-4634-8205-38d53170d5a7"

      state: >

          {% set dir = states('sensor.3d_printer_current_print') %}

          {% set img = states('sensor.3d_printer_object_thumbnails_path') %}

          {{ [dir.split('/')[0], img] | join('/') }}

      attributes:

        friendly_name: "Object Thumbnail"

        icon: mdi:image
```




I guess it could be made with only one sensor but that worked for me.



I also found out that if the filename as a special character in it, such as a  + sign, the api calls using the filename as an argument like in the metadata call would fail because this character wouldn’t be escaped in the url so I modified the api call this way by encoding it with urlencode.


''' yaml

rest:

  scan_interval: 5

  resource_template:  <http://192.168.168.29:7125/server/files/metadata?filename=%7b%7b%20states(('sensor.3d_printer_current_print'))%20|urlencode%20%7d%7d> http://192.168.168.29:7125/server/files/metadata?filename={{ states(('sensor.3d_printer_current_print')) |urlencode }}

  sensor:

    - name: printer_3d_file_metadata

      unique_id: "192.168.168.295ec44d90-419c-419c-802d-d34071639c08"

      json_attributes_path: "$.result"

      json_attributes:

        - layer_height

        - object_height

        - thumbnails

      value_template: "OK"
'''
 

For the layer count, I was looking at the metadata and the current layer was always null therefore your current layer sensor is based on the height and position of the z. I investigated a bit further why the metadata wasn’t showing in the api call and found out there are some GCode to be added to the slicer info. 



Basically for getting the current layer and total_layer to be available as metadata, this needs to be added to the slicer:

in the Start GCode

SET_PRINT_STATS_INFO TOTAL_LAYER=[total_layer_count]

and in the “Before layer change GCode
SET_PRINT_STATS_INFO CURRENT_LAYER=[layer_num]

then on the API call you will get the info of each layer and the exact total layer and then could have a sensor which would report the exact current_layer and total_layer with out any calculation.

**Comment by davisgoodman** [comment]
Just to add more info,  here is the example of the API call 

http:// <http://%3cmoonraker_ip%3e:7125/printer/objects/query?heater_bed&extruder&print_stats&toolhead&display_status&virtual_sdcard&gcode_move&filament_switch_sensor%20Filament&temperature_sensor%20Chamber_Temp%22> <moonraker_ip>:7125/printer/objects/query?heater_bed&extruder&print_stats&toolhead&display_status&virtual_sdcard&gcode_move&filament_switch_sensor%20Filament&temperature_sensor%20Chamber_Temp%22

{"result": {"eventtime": 52339.529589617, "status": {"heater_bed": {"temperature": 23.03, "target": 0.0, "power": 0.0}, "extruder": {"temperature": 20.8, "target": 0.0, "power": 0.0, "can_extrude": false, "pressure_advance": 0.45, "smooth_time": 0.04}, "print_stats": {"filename": "Voron/Tools-Anycubic Mega X-Sparta 3D ABS+.gcode", "total_duration": 16.203441972997098, "print_duration": 0.0, "filament_used": 0.0, "state": "printing", "message": "", "info": {"total_layer": 50, "current_layer": 0}}, "toolhead": {"homed_axes": "xyz", "axis_minimum": [-5.0, 0.0, -5.0, 0.0], "axis_maximum": [300.0, 300.0, 288.0, 0.0], "print_time": 52330.888551087286, "stalls": 0, "estimated_print_time": 52330.68505, "extruder": "extruder", "position": [-5.0, 0.0, 22.597679316888048, 38093.440720000064], "max_velocity": 300.0, "max_accel": 1250.0, "max_accel_to_decel": 1900.0, "square_corner_velocity": 5.0}, "display_status": {"progress": 0.0, "message": "Homing..."}, "virtual_sdcard": {"file_path": "/home/pi/gcode_files/Voron/Tools-Anycubic Mega X-Sparta 3D ABS+.gcode", "progress": 0.01012596951379232, "is_active": true, "file_position": 31042, "file_size": 3065583}, "gcode_move": {"speed_factor": 1.0, "speed": 3600.0, "extrude_factor": 1.0, "absolute_coordinates": true, "absolute_extrude": true, "homing_origin": [0.0, 0.0, 0.0, 0.0], "position": [-5.0, 450.0, 22.610179316888047, 38093.440720000064], "gcode_position": [-5.0, 450.0, 22.610179316888047, -20.0]}, "filament_switch_sensor Filament": {"filament_detected": true, "enabled": true}, "temperature_sensor Chamber_Temp\"": {}}}}

 

And the call for the metadata for the same file Voron/Tools-Anycubic Mega X-Sparta 3D ABS+.gcode where the thumbnail path is missing “Voron/”



{"result": {"size": 3065583, "modified": 1675395760.9446192, "uuid": "e0921488-f2a0-4248-88b5-974166704c41", "slicer": "SuperSlicer", "slicer_version": "2.5.59", "gcode_start_byte": 30633, "gcode_end_byte": 3049974, "layer_count": 50, "object_height": 10.2, "estimated_time": 6163, "nozzle_diameter": 0.4, "layer_height": 0.2, "first_layer_height": 0.25, "first_layer_extr_temp": 245.0, "first_layer_bed_temp": 105.0, "chamber_temp": 46.0, "filament_name": "Sparta 3D ABS+", "filament_type": "ABS", "filament_total": 9082.81, "filament_weight_total": 22.5, "thumbnails": [{"width": 32, "height": 32, "size": 702, "relative_path": ".thumbs/Tools-Anycubic Mega X-Sparta 3D ABS+-32x32.png"}, {"width": 400, "height": 300, "size": 11900, "relative_path": ".thumbs/Tools-Anycubic Mega X-Sparta 3D ABS+-400x300.png"}], "print_start_time": 1675440384.953702, "job_id": "0002BB", "filename": "Voron/Tools-Anycubic Mega X-Sparta 3D ABS+.gcode"}}

**Comment by davisgoodman** [comment]
making sure the code get parsed right

``` yaml
    - name: 3d_printer_object_thumbnails_path
      unique_id: "192.168.168.293988e9be-c910-4634-8205-38d53170d5a6"
      state: '{{ states.sensor.printer_3d_file_metadata.attributes["thumbnails"][1]["relative_path"] }}' 
      attributes:
        friendly_name: "Object Thumbnails Path"
        
    - name: 3d_printer_thumbnail
      unique_id: "192.168.168.293988e9be-c910-4634-8205-38d53170d5a7"
      state: >
          {% set dir = states('sensor.3d_printer_current_print') %}
          {% set img = states('sensor.3d_printer_object_thumbnails_path') %}
          {{ [dir.split('/')[0], img] | join('/') }}
      attributes:
        friendly_name: "Object Thumbnail"
        icon: mdi:image
```

**Comment by davisgoodman** [comment]
And this one where I modified the rest call to be able to handle special characters in the filename
```yaml
  - platform: rest
    scan_interval: 15
    name: klipper_preview_path
    unique_id: "192.168.168.29fd0ae36c-0d51-4392-a18d-89861d536ba4"
    resource_template: "http://192.168.168.29:7125/server/files/metadata?filename={{ states(('sensor.3d_printer_current_print')) |urlencode}}"
    json_attributes_path: "$.result.thumbnails.[2]"
    json_attributes:
      - relative_path
      - width
      - height
      - size
    value_template: "OK"
    
rest:
  scan_interval: 5
  resource_template: "http://192.168.168.29:7125/server/files/metadata?filename={{ states(('sensor.3d_printer_current_print')) |urlencode }}"
  sensor:
    - name: printer_3d_file_metadata
      unique_id: "192.168.168.295ec44d90-419c-419c-802d-d34071639c08"
      json_attributes_path: "$.result"
      json_attributes:
        - layer_height
        - object_height
        - thumbnails
      value_template: "OK"
```

**Comment by NonaSuomy** [comment]
> 
Took a bit but to access three layers deep array you have to do attributes.print_stats["info"]["current_layer"]

Under Home Assistant I played with the Developer Tools -> Template:

```json
{% set value_json=
  {
    "result": 
      {
        "eventtime": 1495044.184335206, 
        "status": 
        {
          "heater_bed": 
          {
            "temperature": 40.3, 
            "target": 0.0, 
            "power": 0.0
          }, 
          "extruder": 
          {
            "temperature": 24.87, 
            "target": 0.0, 
            "power": 0.0, 
            "can_extrude": false, 
            "pressure_advance": 0.0, 
            "smooth_time": 0.04
          }, 
          "print_stats": 
          {
            "filename": "SpringNutxyz002.gcode", 
            "total_duration": 83.25961575005203, 
            "print_duration": 0.0, 
            "filament_used": 0.0, 
            "state": "cancelled", 
            "message": "", 
            "info": 
            {
              "total_layer": 26, 
              "current_layer": 0
            }
          }, 
          "toolhead": 
          {
            "homed_axes": "", 
            "axis_minimum": [0.0, 0.0, -1.5, 0.0], 
            "axis_maximum": [117.5, 118.0, 120.0, 0.0], 
            "print_time": 1776.681568, 
            "stalls": 0, 
            "estimated_print_time": 2041.05257603125, 
            "extruder": "extruder", 
            "position": [0.0, 0.0, 0.0, 0.0], 
            "max_velocity": 200.0, 
            "max_accel": 2000.0, 
            "max_accel_to_decel": 1000.0, 
            "square_corner_velocity": 6.0
          }, 
          "display_status": 
          {
            "progress": 0.0, 
            "message": null
          }, 
          "virtual_sdcard": 
          {
            "file_path": null, 
            "progress": 0.0, 
            "is_active": false, 
            "file_position": 0.0, 
            "file_size": 0.0
          }, 
          "gcode_move": 
          {
            "speed_factor": 1.0, 
            "speed": 1500.0, 
            "extrude_factor": 1.0, 
            "absolute_coordinates": true, 
            "absolute_extrude": true, 
            "homing_origin": [0.0, 0.0, 0.0, 0.0], 
            "position": [0.0, 0.0, 0.0, 0.0], 
            "gcode_position": [0.0, 0.0, 0.0, 0.0]
          }, 
          "filament_motion_sensor btt_smart_filament_sensor": 
          {
            "filament_detected": true, 
            "enabled": true
          }, 
          "temperature_sensor Chamber_Temp": 
          {
            "temperature": 25.67, 
            "measured_min_temp": 25.46, 
            "measured_max_temp": 27.06
          }
        }
      }
    }
%}
{{value_json.result.status.print_stats["info"]["total_layer"]}}

Result type: number
26
```
I combined both values into one button for fun as I was going to do that with the other two layer buttons.

Which should then look like this:

![image](https://user-images.githubusercontent.com/1906575/216767877-7be4bb6e-d02a-43c2-801c-ecca1c55953f.png)

```yaml
    - name: 3d_printer_actual_layer
      unique_id: "IP167e5dec-719c-42d3-9560-4f177573f742"
      state: '{{ states.sensor.printer_3d_sensor.attributes.print_stats["info"]["current_layer"] }} / {{ states.sensor.printer_3d_sensor.attributes.print_stats["info"]["total_layer"] }}'
      availability: >
        {% set items = ['sensor.printer_3d_sensor'] %}
        {{ expand(items)|rejectattr('state','in',['unknown','unavailable'])
          |list|count == items|count }}
      icon: mdi:counter
      attributes:
        friendly_name: "Actual Layer"
```

**Comment by davisgoodman** [comment]
> ```yaml
>     - name: 3d_printer_actual_layer
>       unique_id: "IP167e5dec-719c-42d3-9560-4f177573f742"
>       state: '{{ states.sensor.printer_3d_sensor.attributes.print_stats["info"]["current_layer"] }} / {{ states.sensor.printer_3d_sensor.attributes.print_stats["info"]["total_layer"] }}'
>       availability: >
>         {% set items = ['sensor.printer_3d_sensor'] %}
>         {{ expand(items)|rejectattr('state','in',['unknown','unavailable'])
>           |list|count == items|count }}
>       icon: mdi:counter
>       attributes:
>         friendly_name: "Actual Layer"
> ```

That's great! combines both at the same time!

**Comment by NonaSuomy** [comment]
> making sure the code get parsed right
> 
> ```yaml
>     - name: 3d_printer_object_thumbnails_path
>       unique_id: "192.168.168.293988e9be-c910-4634-8205-38d53170d5a6"
>       state: '{{ states.sensor.printer_3d_file_metadata.attributes["thumbnails"][1]["relative_path"] }}' 
>       attributes:
>         friendly_name: "Object Thumbnails Path"
>         
>     - name: 3d_printer_thumbnail
>       unique_id: "192.168.168.293988e9be-c910-4634-8205-38d53170d5a7"
>       state: >
>           {% set dir = states('sensor.3d_printer_current_print') %}
>           {% set img = states('sensor.3d_printer_object_thumbnails_path') %}
>           {{ [dir.split('/')[0], img] | join('/') }}
>       attributes:
>         friendly_name: "Object Thumbnail"
>         icon: mdi:image
> ```

Did you test if this works recursively to multiple directories deep? /directory1/directory2/directory3/.thumbs/test.png etc?

**Comment by NonaSuomy** [comment]
Discussion on HA Discord petro says to use 
`[:-1]` if you want everything before the last 
everything after the first `dir.split('/')[1:]` 
everything before the last 2.. `dir.split('/')[:-2]`

so...

```
{{ [dir.split('/')[:-1], img] | join('/') }}
```
to remove the remaining directory from the end.

**Comment by NonaSuomy** [comment]
Multi and no directory fix for thumbnails.
```
     - name: 3d_printer_object_thumbnails
       unique_id: "IP59b37837-b751-4d31-98c2-516a52edf833"
       state: >
         {% set dir = states('sensor.3d_printer_current_print') %}
         {% set img = states.sensor.printer_3d_file_metadata.attributes["thumbnails"][2]["relative_path"] %}
         {{ (dir.split('/')[:-1] + [img]) | join('/') }}
       availability: >
         {% set items = ['sensor.printer_3d_file_metadata'] %}
         {{ expand(items)|rejectattr('state','in',['unknown','unavailable'])
           |list|count == items|count }}
       icon: mdi:image
       attributes:
         friendly_name: "Object Thumbnails"
```

**Comment by davisgoodman** [comment]
@NonaSuomy 
So for full directory parsing whether there is one or 10 subdirectories, this code will work, but you need to change the camera sensor to the 3d_printer_object_thumbnails instead of 3d_printer_thumbnails

```yaml

    - name: 3d_printer_object_thumbnails
      unique_id: "IP59b37837-b751-4d31-98c2-516a52edf833"
      state: >
        {% set dir = states('sensor.3d_printer_current_print') %}
        {% set img = state_attr('sensor.printer_3d_file_metadata', 'thumbnails')[1].relative_path %}
        {{ (dir.split('/')[:-1] + [img]) | join('/') }}
      availability: >
        {% set items = ['sensor.printer_3d_file_metadata'] %}
        {{ expand(items)|rejectattr('state','in',['unknown','unavailable'])
          |list|count == items|count }}
      icon: mdi:image
      attributes:
        friendly_name: "Object Thumbnails"    
       
 ```

**Comment by davisgoodman** [comment]
@NonaSuomy 
While configuring KlipperScreen it mentioned the same thing about getting Print_Stats_info but in order to have accurate layer position they suggest to add SET_PRINT_STATS_INFO CURRENT_LAYER={layer_num + 1} in the "After Layer Change G-Code.

The reference is here: https://klipperscreen.readthedocs.io/en/latest/Quicktips/#layer-progress

**Comment by NonaSuomy** [comment]
Try the current code it should always select the last thumbnail.
